### ISSUE TITLE
Meta-Gradient through a KL_Div loss is zero

### ISSUE THREAD

**Issue by pratikgujjar** [facebookresearch/higher#64]
I am trying to implement this paper (https://arxiv.org/pdf/2003.10580v2.pdf). The goal here is to jointly learn parameters for a teacher and a student model such that student trained on the the teacher's predictions can perform well on the validation set.

Code to reproduce this:
```
import torch
import higher
from torch import nn
from torch import optim
from torch.nn import functional as F

class Student(nn.Module):
    def __init__(self):
        super(Student, self).__init__()
        w = torch.empty(3, 5)
        self.w = nn.Parameter(nn.init.constant_(w, 3))

    # Defining the forward pass
    def forward(self, x):
        x = self.w * x
        return x

class Teacher(nn.Module):
    def __init__(self):
        super(Teacher, self).__init__()
        w = torch.empty(3, 5)
        self.w = nn.Parameter(nn.init.constant_(w, 3.1))

    # Defining the forward pass
    def forward(self, x):
        x = self.w * x
        return x

teacher = Teacher().cuda()
student = Student().cuda()
mse_loss = torch.nn.MSELoss()
teacher_optim = optim.SGD(teacher.parameters(), lr=0.01)
student_optim = optim.SGD(student.parameters(), lr=0.01)

x_train = torch.ones(3, 5).cuda()
x_val = (torch.ones(3, 5) * 2).cuda()
y_val = (torch.ones(3, 5) * 8).cuda()

with torch.set_grad_enabled(True):
    with higher.innerloop_ctx(student, student_optim, copy_initial_weights=True) as (fmodel, diffopt):
        # Collect pseudo labels for train set
        pseudo_targets = teacher(x_train)

        # Train student on y_pseudo
        logits = fmodel(x_train)
        pseudo_loss = F.kl_div(F.log_softmax(logits, dim=1), F.softmax(pseudo_targets), reduction='batchmean')
        # pseudo_loss = F.mse_loss(logits, pseudo_targets)
        diffopt.step(pseudo_loss)

        # Update teacher based on student performance on validation set
        logits_q = fmodel(x_val)

    meta_loss = F.mse_loss(logits_q, y_val)
    grad_t = torch.autograd.grad(meta_loss, teacher.parameters())
    print (grad_t)
```

The issue I have noticed is when I use KL_Divergence loss for pseudo loss, grad_t is zero. When I revert to the mse_loss, all is fine. For this toy case, mse_loss sounds appropriate, but for when you're working with classification, KL_Div is needed. I have checked that pseudo_targets do indeed require_gradients. 

Does anyone know if I am doing something wrong here?

**Comment by egrefen** [comment]
Your logits and pseudo targets in log space are very close. This means that `F.softmax(F.log_softmax(logits, dim=1) == F.softmax(pseudo_targets)` (not mathematically, but due to precision limitations). As a result, the `kl_loss` function is computing the KL loss over two identically parameterized categorials, which is a loss of zero, which is the minimal value, hence a gradient of zero. When you do the MSE loss, there's a difference between 3 and 3.1 which doesn't get washed out by precision issues, and therefore your loss is non-zero.

This is not a `higher` related issue, but just caused by your data (both pytorch and higher are working as intended), so I'm closing it.

**Comment by pratikgujjar** [comment]
Thank you @egrefen for looking into it. I suppose it is my fault that I whipped this code up as a parallel to the CIFAR10 data I was actually working with, in where, the problem still persists. 

If you allow me to change the models to small CNNs and change the problem to that of a classification with new image-like inputs, the `pseudo_loss` is non-zero. And yet, `grad_t` is zero as long as the `pseudo_loss` is `kl_div` and not `mse_loss`, though the magnitudes of both losses are close.

```
import torch
import higher
from torch import nn
from torch import optim
from torch.nn import functional as F

class Teacher(nn.Module):
    def __init__(self):
        super(Teacher, self).__init__()

        self.cnn_layers = nn.Sequential(
            # Defining a 2D convolution layer
            nn.Conv2d(3, 16, kernel_size=3, stride=1, padding=1),
            nn.ReLU(inplace=True),
            nn.BatchNorm2d(16),
            nn.MaxPool2d(kernel_size=2),
            # Defining another 2D convolution layer
            nn.Conv2d(16, 16, kernel_size=3, stride=1, padding=1),
            nn.BatchNorm2d(16),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2),
            # Defining another 2D convolution layer
            nn.Conv2d(16, 16, kernel_size=3, stride=1, padding=1),
            nn.BatchNorm2d(16),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2))
        self.linear_layers = nn.Sequential(nn.Linear(16 * 4 * 4, 10))

    # Defining the forward pass
    def forward(self, x):
        x = self.cnn_layers(x)
        x = x.view(x.size(0), -1)
        x = self.linear_layers(x)
        return x


class Student(nn.Module):
    def __init__(self):
        super(Student, self).__init__()

        self.cnn_layers = nn.Sequential(
            # Defining a 2D convolution layer
            nn.Conv2d(3, 16, kernel_size=3, stride=1, padding=1),
            nn.BatchNorm2d(16),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2),
            # Defining another 2D convolution layer
            nn.Conv2d(16, 16, kernel_size=3, stride=1, padding=1),
            nn.BatchNorm2d(16),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2))

        self.linear_layers = nn.Sequential(nn.Linear(16 * 8 * 8, 10))

    # Defining the forward pass
    def forward(self, x):
        x = self.cnn_layers(x)
        x = x.view(x.size(0), -1)
        x = self.linear_layers(x)
        return x

teacher = Teacher().cuda()
student = Student().cuda()
mse_loss = torch.nn.MSELoss()
teacher_optim = optim.SGD(teacher.parameters(), lr=0.01)
student_optim = optim.SGD(student.parameters(), lr=0.01)

# Shape of 32x32
x_train = (torch.ones(1, 3, 32, 32) * 0.8).cuda()
x_val = (torch.ones(1, 3, 32, 32) * 0.5).cuda()

# Shape of 1
y_val = torch.ones(1).type(torch.LongTensor).cuda()

with torch.set_grad_enabled(True):
    with higher.innerloop_ctx(student, student_optim, copy_initial_weights=True) as (fmodel, diffopt):
        # Collect pseudo labels for train set
        pseudo_targets = teacher(x_train)

        # Train student on y_pseudo
        logits = fmodel(x_train)
        pseudo_loss = F.kl_div(F.log_softmax(logits, dim=1), F.softmax(pseudo_targets), reduction='batchmean')
        # pseudo_loss = F.mse_loss(logits, pseudo_targets)
        diffopt.step(pseudo_loss)

        # Update teacher based on student performance on validation set
        logits_q = fmodel(x_val)

    meta_loss = F.cross_entropy(logits_q, y_val)
    grad_t = torch.autograd.grad(meta_loss, teacher.parameters())
    print (grad_t)
```
